### ISSUE TITLE
Suggest similarly spelled search results

### ISSUE THREAD

**Issue by mhutti1** [kiwix/kiwix-android#41]
_From @rashiq on June 8, 2016 13:6_

**Reported by rauldelacalle on 2015-03-09 01:16 UTC**
SPANISH;

Buen día.

Descargue la aplicacion Kiwix para android y la instalé. Descargué el Wikcionario en español, lo descomprimí y lo cargue en la memoria externa del telefono inteligente. Desde la aplicación leo el archivo y todo está bien.

Pero cuando escribo una palabra de manera equivocada el Wikcionario no me corrige. Ejemplo: en español la forma correcta es: Zapato. Si escribo sapato la aplicación me dice: Error: "falló la carga del artículo sapato", pero no me corrige debería mostrame las opciones. Es decir que tengo que ser un experto en el idioma para buscar, no me sirve de esa manera pues el objetivo es que me corrija cuando yo estoy equivocado.

Si hago lo mismo en el computador me muestra como opciones:
1. Zapato
2. Calzado
3. Pasta de zapatos
- Es posible mejorar la aplicación para android?
- Estoy fallando en algo?

ENGLISH:

Good day.

Download the Kiwix application for android and I installed it. I downloaded the Wiktionary in Spanish, unzipped it and upload it to the external memory of the smart phone. Since I read the application file and all is well.

But when I write the wrong word Wiktionary does not correct me. Example: in Spanish is written: ZAPATO. If SAPATO write the application tells me: Error: "failed load SAPATO article", but does not correct me should show me the options. You mean I have to be an expert in the language to find, does not help me that way because the objective is to correct me when I'm wrong.

If I do the same on the computer shows me as options:
1. Zapato
2. Calzado
3. Pasta de zapatos
- It is possible to improve the application for android?
- I'm failing at something?

_Copied from original issue: kiwix/kiwix#193_

**Comment by mhutti1** [comment]
_From @rashiq on June 8, 2016 13:14_

**Updated by kelson42 on 2015-06-23 04:55 UTC**
- **labels**:  --> android
- **Group**: v1.0 --> None

**Comment by mhutti1** [comment]
This appears to be a search limitation. @kelson42 what is the desktop version do, that the app doesn't that makes search more intelligent?

**Comment by mhutti1** [comment]
_From @Frenzie on March 2, 2017 14:32_

For search query results it would be nice to make them ignore case, diacritics, and possibly punctuation (except that the matches closest to what you typed should come higher in the results list). Especially on a phone it's much easier to type `deja vu` or `deja-vu` than `déjà-vu`. Basically like Aard Dictionary.

**Comment by srv-twry** [comment]
@mhutti1 I'll give it a shot!

**Comment by kelson42** [comment]
Probably something which needs to be delegated to the kiwix-lib/Xapian. I strongly suspect there is nothing much we should do on Java side.

**Comment by rashiq** [comment]
We could probably play around with [levenshtein distance](https://en.wikipedia.org/wiki/Levenshtein_distance)

**Comment by kelson42** [comment]
All of this should be done in the kiwix-lib anyway. The libzim has already a Levenshtein distance implementation.

**Comment by kelson42** [comment]
Will generate a new WPES ZIM file and check this again.

**Comment by rashiq** [comment]
Gotcha

**Comment by Frenzie** [comment]
@kelson42 The latest Kiwix for Android on F-Droid and Google Play (2.2) is from almost exactly one year ago, 2017-03-21, so unsurprisingly when you type `deja vu` it doesn't find `déjà-vu`. 

Is there some kind of nightly build easily available to test with or would one have to build it?

**Comment by kelson42** [comment]
All the problems around diacritics or case should be fixed, but we still don’t have a solution to propose similar words. This is wha this tickets depends on https://github.com/openzim/libzim/issues/731

**Comment by kelson42** [comment]
@MohitMaliFtechiz 

Here is the refinement of this very old issue:
* If something is typed but does not lead to any suggestion (search result), then Kiwix should try to propose an altnernative similar word (kind of assuming there is a typo in the word)
* The libzim propose a primitive (still to be released) allowing to deliver an altnernative workd given one (so something like `string alternative spellcheck(string original)`
* Once an altnernative spelling get from the libzim, Kiwix should display the altnernative below the search box with the following sentence "Do you mean **xyz**?" ("xyz" been the altnernative spelling of the search pattnern)
* If the user click on **xyz** then Kiwix should replace current search with "xyz" and run immediatly the suggestions

You can also follow the Apple version of this feature request, see https://github.com/kiwix/kiwix-apple/issues/1035 tha we implement right now

Can you please implement a mockup ASAP (even if the libizm is still not over) so we can validate with DWDS which needs urgently this feature?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 Here is a video sample of this feature.

https://github.com/user-attachments/assets/035fcc67-41cb-4ea8-a2e7-13131f912e3c

**Comment by kelson42** [comment]
@gremid How does it look like to you?
@MohitMaliFtechiz Looks good to me, although I'm not sure people will understand they can click on it. What about adding the link icon like for the suggestions? Woukd that help?

**Comment by MohitMaliFtechiz** [comment]
@kelson42 You seems right here, adding link will give the hint to user that this is a clickable link. It will look like this.

<img width="1080" height="2400" alt="Image" src="https://github.com/user-attachments/assets/dc60ae21-0941-4520-8905-8e6bcb6bc421" />

**Comment by gremid** [comment]
I actually prefer the unified solution proposed for iOS: https://github.com/kiwix/kiwix-apple/issues/1035#issuecomment-3365205364

The use case is comparable to that of a typical "auto-complete" widget. It retrieves potential matches from an index, prioritizing exact matches over fuzzy ones. We could flag suggested matches (i.e., with an icon or other visual cues) so that the user knows the match type.  Switching between lists suddenly, and also changing the visual alignment of the entries because of the prefix "Do you mean?" seems flaky to me. It also needs to get localized, adding another layer of complexity. 

IMHO, users are used to auto-complete widgets being exact as well as fuzzy.

**Comment by Frenzie** [comment]
Agreed, there could be a separator or some kind of icon indicating it's not an exact match but the "do you mean" part strikes me as unnecessarily complicated and distracting.

**Comment by gremid** [comment]
Sorry, I watched the iOS solution again and noticed that suggestions are also provided in a separate drop-down. As I commented there and to reiterate my previous point, please do not provide exact and fuzzy matches in separate widgets/lists.

**Comment by kelson42** [comment]
> Sorry, I watched the iOS solution again and noticed that suggestions are also provided in a separate drop-down. As I commented there and to reiterate my previous point, please do not provide exact and fuzzy matches in separate widgets/lists. 

It seems we agree on this, this is basically my request on previous comment. New widget system is not bad in itself, but here it brings confusion.
